scaffoldmaker/utils/interpolation.py

```diff
@@ -787,3 +787,61 @@ def getDoubleCubicHermiteCurvesMidDerivative(ax, ad1, mx, bx, bd1):
     magb = vector.magnitude(bd1)
     magm = arcLengtha + arcLengthb - 0.5*(maga + magb)
     return vector.setMagnitude(md1, magm)
+
+def sampleParameterAlongCenterLine(paramList, lengthList, outputElementsCount):
+    """
+    Generates cubic hermite curve using parameter values in paramList at each
+    location along on the center line as defined in lengthList. The function
+    generates parameter value and derivative at equally spaced points along the
+    center line.
+    :param paramList: List of parameter values at length locations specified
+    in lengthList.
+    :param lengthList: List of length locations along center line.
+    :param outputElementsCount: Number of output elements along total length.
+    :return paramList, dParamList: Parameter values and rate of change at each
+    sampled point.
+    """
+
+    nx = []
+    nd1 = []
+    xList = []
+    d1List = []
+    for i in range(len(lengthList)):
+        v = [lengthList[i], paramList[i]]
+        d1 = [(lengthList[i+1] - lengthList[i]) if i < len(lengthList)-1 else (lengthList[i] - lengthList[i-1]), 0.0]
+        nx.append(v)
+        nd1.append(d1)
+    smoothd1 = smoothCubicHermiteDerivativesLine(nx, nd1)
+
+    # Get total arclength
```

Comment:
I think this should be sampling at equal distances along the lengths, not including the parameter as a component of coordinates (it may after all be used for a non-spatial value). It would look increasingly wrong if there was a lot of variation of the parameter at one end vs. the other.
I think we need a special smoothing function to get derivatives of parameter p with linearly varying x lengths.
The arc length should just be the total length of the line (last x - first x). Find xi in the original elements at sample lengths, then evaluate p, dp/dxi, dx/dxi and use chain rule to get dp/dx. Divide by outputElementsCount to get derivatives for the segment length?
Probably easier to describe on paper.